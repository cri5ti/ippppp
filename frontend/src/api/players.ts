import {API_URL} from "../config";


export type TPlayer = {
  code: string;
  description: string;
  email: string;
  wins: number;
  losses: number;
}

export const playersApi = {
  getAll: async (): Promise<TPlayer[]> => await (await fetch(API_URL + "/players")).json(),
  getOne: async (code: string): Promise<TPlayer> => await (await fetch(`${API_URL}/players/${code}`)).json(),
  create: async (body: any): Promise<Response> => await fetch(`${API_URL}/players`, {
    method: "PUT",
    body: JSON.stringify(body),
    headers: {
      'Accept': 'application/json',
      'Content-Type': 'application/json'
    }}),
  remove: async (code: string): Promise<Response> => await fetch(`${API_URL}/players/${code}`, {method: "DELETE"})
};
