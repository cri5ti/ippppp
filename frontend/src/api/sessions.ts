import {API_URL} from "../config";

export type Session = {
    code: string,
    description: string,
    minGamesRequired: number,
    isActive: boolean,
    sessionPlayers: Array<any> //todo sessionPlayer
}

export const sessionApi = {
    getAll: async (): Promise<Session[]> => await (await fetch(API_URL + "/sessions")).json(),
    getOne: async (code: string): Promise<Session> => await (await fetch(`${API_URL}/sessions/${code}`)).json(),
    create: async (body: Partial<Session>): Promise<Response> => await fetch(`${API_URL}/sessions`, {
        method: "PUT",
        body: JSON.stringify(body),
        headers: {
            'Accept': 'application/json',
            'Content-Type': 'application/json'
        }}),
    remove: async (code: string): Promise<Response> => await fetch(`${API_URL}/sessions/${code}`, {method: "DELETE"})
};
