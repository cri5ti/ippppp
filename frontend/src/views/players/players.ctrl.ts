class PlayersCtrl{

    public getAll = async (baseUrl?: string): Promise<any> => {
        return await (await fetch(baseUrl ? `${baseUrl}/api/players` : `/api/players`)).json();
    }

}


export const playersCtrl = new PlayersCtrl();
