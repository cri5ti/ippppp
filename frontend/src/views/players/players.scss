.players {
  display: grid;

  grid-template-rows: 1fr auto;

  .list {
    display: grid;
    grid-template-columns: repeat(3, 1fr);
    grid-gap: 0;

    li {
      $c: #0002;
      border-top: solid 1px $c;
      border-left: solid 1px $c;
      &:nth-child(3n+3) { border-right: solid 1px $c; }
      &:nth-last-child(-n+3) { border-bottom: solid 1px $c; }
      &:last-child { border-right: solid 1px $c; }
    }
  }

  .form{
    padding: 20px;
    display: flex;
    flex-direction: column;
  }
}

.player {
  display: grid;
  grid-template-columns: auto 1fr;
  align-items: center;
  gap: 1em;

  img { border-radius: 100%; padding: 1em; }

  .name {

  }
}
