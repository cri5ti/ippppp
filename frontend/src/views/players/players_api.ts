import {API_URL} from "../../config";


export type Player = {
  description: string;
  email: string;
  code: string;
}

export const playersApi = {
  getAll: async (): Promise<Player[]> => await (await fetch(API_URL + "/players")).json()
};
