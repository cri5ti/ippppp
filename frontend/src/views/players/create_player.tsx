import React, {useCallback, useState} from "react";
import {withRouter} from "react-router";
import {BackLink, DefaultButton} from "../../ui/back_button";
import {Page} from "../shell/shell";
import {playersApi} from "../../api/players";
import {BusyOverlay} from "../../ui/busy/busy";

export const CreatePlayer = withRouter(({history}) => {
    const [form, setForm] = useState<{ name: string, email: string }>({name: null, email: null});
    const [busy, setBusy] = useState(false);

    const onSubmit = useCallback(async (ev) => {
        ev.preventDefault();
        setBusy(true);
        await playersApi.create({Email: form.email, Description: form.name});
        setBusy(false);
        history.replace('/players');
    }, [form]);

    return (
        <Page title="New player">
            <nav>
                <DefaultButton onClick={onSubmit}>Create</DefaultButton>
                <BackLink/>
            </nav>

            <form className="form">
                {busy && <BusyOverlay/>}

                <label htmlFor="name">Name:</label>
                <input required={true} type="text" id="name" name="name" value={form.name}
                       onChange={(e) => setForm({...form, name: e.target.value})}/>

                <label htmlFor="email">Email:</label>
                <input required={true} type="email" id="email" title="Email" value={form.email}
                       onChange={(e) => setForm({...form, email: e.target.value})}/>
            </form>
        </Page>
    )
});
