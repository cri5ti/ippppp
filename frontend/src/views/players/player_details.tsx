import React, {useEffect, useState} from "react";
import {useHistory, useParams} from "react-router";
import {BackLink, DefaultButton} from "../../ui/back_button";
import {BusyOverlay} from "../../ui/busy/busy";
import {Page} from "../shell/shell";
import {Player, playersApi} from "./players_api";

export const PlayerDetails = () => {
    const {code} = useParams();
    const [player, setPlayer] = useState<Player>(null);

    const history = useHistory();

    useEffect(() => {
        async function load() {
            const player = await playersApi.getOne(code);
            setPlayer(player);
        }
        load();
    }, []);

    async function onDelete() {
        const res = await playersApi.remove(player.code);
        if(!res) return;
        history.goBack();
    }

    if (!player) return <BusyOverlay/>;

    return (
        <Page title={player && player.description}>
            <nav>
                <BackLink/>
                <DefaultButton onClick={onDelete}>Delete</DefaultButton>
            </nav>

            {!player && <BusyOverlay/>}

            {player && <>
                {player.code}
            </>}

            <p>Details go here</p>
        </Page>
    );
};
