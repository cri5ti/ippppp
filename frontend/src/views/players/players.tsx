import React from "react";
import {Link, Route, Switch, useRouteMatch, withRouter} from "react-router-dom";
import {List} from "../../ui/list/list";
import {md5} from "../../util/md5";
import {Page} from "../shell/shell";
import {CreatePlayer} from "./create_player";
import {PlayerDetails} from "./player_details";
import {Player, playersApi} from "./players_api";

const css = require('./players.scss');

const PlayersPage = () => {
    const {path} = useRouteMatch();

    return (
        <Switch>
            <Route path={`${path}/create`}>
                <CreatePlayer/>
            </Route>
            <Route path={`${path}/:code`}>
                <PlayerDetails/>
            </Route>
            <Route>
                <PlayersList/>
            </Route>
        </Switch>
    );
};


const PlayersList = withRouter(({history}) => {
    const {url} = useRouteMatch();

    return (
        <Page title="Players">
            <nav>
                <Link to={`${url}/create`}>Create player</Link>
            </nav>

            <div className="players">
                <List itemRender={player => <PlayerItem {...player}/>}
                      data={playersApi.getAll()}
                      onItemClick={(i) => history.push(url + '/' + i.code)}/>
            </div>
        </Page>
    )
});


const PlayerItem = (props: Player) => {
    const {description, email} = props;

    return (
        <div className="player">
            <Gravatar email={email} size={64}/>
            <span className="name">{description}</span>
        </div>
    )
};

const Gravatar = ({email, size, ...rest}) => <img src={`https://secure.gravatar.com/avatar/${md5(email)}?size=${size}`} {...rest}/>

export default PlayersPage;
