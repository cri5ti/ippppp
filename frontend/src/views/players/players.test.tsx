import React from "react";
import {render, unmountComponentAtNode} from "react-dom";
import {act} from "react-dom/test-utils";
import {BrowserRouter as Router} from 'react-router-dom';
import PlayersPage from "./players";
import {playersApi} from "../../api/players";

jest.mock('./players_api', () => ({
  playersApi: {
    getAll: jest.fn(() => [
      {code:'p1', description: "John", email: 'john@example.com'},
      {code:'p2', description: "Doe", email: 'doe@example.com'}
    ])
  }
}));

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useRouteMatch: () => ({url: '/players'}),
}));

let container;

beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe("Players", () => {
  it('snapshot testing', async () => {
    await act(async () => {
      render(<Router><PlayersPage /></Router>, container);
    });

    expect(container).toMatchSnapshot();
  });
});
