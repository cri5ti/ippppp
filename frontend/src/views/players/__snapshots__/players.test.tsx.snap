// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Players snapshot testing 1`] = `
<div>
  <div
    class="page"
  >
    <header>
      Players
    </header>
    <main>
      <nav>
        <a
          href="/players/create"
        >
          Create player
        </a>
      </nav>
      <div
        class="players"
      >
        <ol
          class="list"
        >
          <li>
            <div
              class="player"
            >
              <img
                src="https://secure.gravatar.com/avatar/d4c74594d841139328695756648b6bd6?size=64"
              />
              <span
                class="name"
              >
                John
              </span>
            </div>
          </li>
          <li>
            <div
              class="player"
            >
              <img
                src="https://secure.gravatar.com/avatar/de21b8c123847c80205e93b301437b45?size=64"
              />
              <span
                class="name"
              >
                Doe
              </span>
            </div>
          </li>
        </ol>
      </div>
    </main>
  </div>
</div>
`;
